### PR TITLE
Revised Battery Monitoring

### PR DESCRIPTION
**Pull request by orangelynx**
Okay so finally here we go..

This PR is regarding https://github.com/PX4/Firmware/issues/2529

I have spent quite some time about the best way to tackle this problem, especially the one that arises from voltage drop of older batteries. I think that the solution I came up with is the best deal between comfort for inexperienced users and experienced users who want to get the last mah out of their batteries.

First off, the flight pilot has (so far) no need to know what kind of battery is powering the system, as  we do not make different calculations or assumptions on this information in the code. Therefore the FC SHOULDN'T make any assumption either. It may be a LiPo powering the system, a car battery or  even a power lead. Therefore I removed the `BAT_N_CELLS` and `BAT_V_LOADDROP` parameters.

I chose the remaining battery parameters with the following thought in my mind: The parameters should directly be backed by functionality, i.e. the actual algo of the battery monitoring system. Any helping parameter "translations" can be done by the GroundStation, because an inexperienced user will usually rely on the setup wizard instead of changing parameters directly. The experienced user however will be able to directly tweak the monitoring system using the given parameters.

The new parameters used for battery monitoring are:
`BAT_V_FULL` - The total full battery voltage.
`BAT_V_LOW` - The voltage at which the low battery alarm will be triggered.
`BAT_V_CRIT` - The voltage at which the critical battery alarm will be triggered.
`BAT_CAPACITY` - The total capacity of the battery.
`BAT_PERC_LOW` - The percentage capacity left at which the low battery alarm will be triggered.
`BAT_PERC_CRIT` - The percentage capacity left at which the critical battery alarm will be triggered.

This might seem confusing at first.. why absolute voltage levels for low and critical but percentage values for capacity? I will come back to that during my explanation of the battery monitoring algorithm. The main advantage of these parameters are, that an experienced users who knows what his batteries can handle, can directly include the max. voltage drop under load into the `BAT_V_LOW` and `BAT_V_CRIT` parameters, thus saving the hazzle of implementing complex, resource intensive and probably not very robust models of the battery (which would again rely on the type of battery in use).

Inexperienced users should be guided through the setup process by the GS. If they run into problems, they should be directed to a battery guide  explaining how to setup the parameters directly.

Now to the actual battery monitoring system. During coding the algo, I came across two variants, and both are included in the source code (one as comment).
In any case, we want to rely on coulomb-counting, as it will give us a better estimate (as it doesn't suffer from something like drop under load). However if the battery voltage enters critical levels, we want those estimations to "take over".

The coulomb-counting estimation is pretty straight forward and is not elaborated further.

```C
	/* VOLTAGE BASED ESTIMATION - Optimistic Voltage / Discrete Approximation */
	if (voltage < bat_v_crit) {
		bat_status_v = bat_perc_crit;
	} else if (voltage < bat_v_low)	{
		bat_status_v = bat_perc_low;
	} else {
		bat_status_v = 1.0f;
	}
```

The voltage estimation "in use" is a discrete "optimistic" voltage estimation. This means, that the current voltage is compared against the low and critical thresholds and will then set the `battery_remaining` variable to the according percentage threshold. A "pessimistic" estimation would set the estimation to just above the next threshold (i.e. for critical instead to `bat_perc_crit` to 0)

The other possible voltage estimation is the linear approximation. In this case, the `voltage` will be mapped to the `bat_remaining` value like this:
(`bat_v_full`, `bat_v_low`) -> (1.0, `bat_perc_low`)
(`bat_v_low`, `bat_v_crit`) -> (`bat_perc_low`, `bat_perc_crit`)
(`bat_v_crit`, 0.0) -> (`bat_perc_crit`, 0.0)

here, the pessimistic case would map the last mapping to constant zero, as technically `bat_v_critical` is already fully empty. so the algo has a little flaw here.

Below you will find a few graphs illustrating the behavior.

So which of the two should be preferred? This is something that the userbase will have to decide, I personally prefer the discrete version, because it "interferes" less with the coulomb-counting approach.

Neither of the two have been fully flight tested, however a slightly modified version of the discrete one has, and both should work fine.

If anybody is interested in the MATLAB script simulation that creates the graph below (the simulation should be clearly distinguishable from the real data plots), I can post this as well.

That is it for now, I hope I didn't forget anything..

EDIT 1: I'm aware that the code is not optimized in any way, that is on purpose. First of, I hope the optimizer will do its work and secondly, this is much better for illustrating the algo.

batmonv
batmonc
batmonsim

### CODE REVIEW

File: src/modules/commander/commander_params.c [hunk 1]
```diff
@@ -87,58 +87,73 @@ PARAM_DEFINE_FLOAT(TRIM_PITCH, 0.0f);
...
+ * Battery capacity.
  *
- * Defines the number of cells the attached battery consists of.
+ * Defines the capacity of the attached battery.
  *
  * @group Battery Calibration
- * @unit S
- * @min 1
- * @max 10
+ * @unit mAh
```
Comment (DonLakeFlyer) [inline src/modules/commander/commander_params.c]: Seems odd to be a float instead of an int. Also I assume -1 signals not set? If so, can you put that in the description?

File: src/modules/commander/commander_params.c [hunk 2]
```diff
@@ -87,58 +87,73 @@ PARAM_DEFINE_FLOAT(TRIM_PITCH, 0.0f);
...
+ * Defines the percentage charge at which the battery is considered low.
  *
  * @group Battery Calibration
- * @unit mA
+ * @unit %
+ * @min 0.0
+ * @max 100.0
  */
-PARAM_DEFINE_FLOAT(BAT_CAPACITY, -1.0f);
+PARAM_DEFINE_FLOAT(BAT_PERC_LOW, 20.0f);
```
Comment (DonLakeFlyer) [inline src/modules/commander/commander_params.c]: Even though this isn't specific to LiPo's. It will probably be lipos most of the time. Going below 30% capacity on a LiPo is usually a bad idea. So wonder about starting at 30% instead of 20%.

**Comment by DonLakeFlyer** [comment]
Although number of cells doesn't make sense in all cases. It allows you to put in per-cell voltage defaults which will work for lipos as good defaults (since that is a very high use case). And then the user can just update cell counts and still use the voltage defaults as is. If your using something odd, you can just set cell count to 1 and adjust all the values yourself.

Also it may make sense to trigger a required setup in QGC. So for example, you could set cell count and/or capacity to default values which indicate not yet set. Then QGC uses that as a trigger for the Power config requiring them to be changed by the user to valid values before first flight.

File: src/modules/commander/commander_params.c (hunk 1 above)
Comment (orangelynx) [inline src/modules/commander/commander_params.c]: yup, well spotted

File: src/modules/commander/commander_params.c (hunk 2 above)
Comment (orangelynx) [inline src/modules/commander/commander_params.c]: again yup, i only know about the magic 80% that one can draw from a LiPo and therefore probably this should be critical.

**Comment by orangelynx** [comment]
I see your point, however since it is of no interest to the actual monitoring system I would put the "shadowing" parameters cell count etc to the battery setup of qgroundcontrol, which then translates it to the actual parameters. Also I don't think that it's very common to use batteries with different cell counts in the same UAV.

**Comment by DonLakeFlyer** [comment]
> however since it is of no interest to the actual monitoring system

I think it's of important interest to the flight stack. Good/Bad parameter defaults affect the usability of the overall system. The reason being, most people don't know they need to change defaults in order for the right thing to happen. Since the voltages are defaulted to specific values it will work incorrectly for a decent percentage of users. Unless they realize they need to change the defaults. QGC is not always used. The firmware should be able to stand alone and do the right thing as much as possible.

I'd suggest either one of these:
- Use a cell count parameter which is not set as a trigger and then voltages which are appropriate for a single lipo cell.
- Set voltages to "not set" by default.

That way you are not getting strange results by accident. And then QGC can also use these are triggers for required setup on Power config.

**Comment by orangelynx** [comment]
Right, the second option is basically how i intended it to be. My main part was to think of a suitable monitoring algo, I didn't yet think about how to achieve good default behaviour.

Point stilll is, that - either way - the firmware has no direct advantage of knowing the cell count, it just adds more terms to the equation. However sensible default values are of course required and since it would be as fatal to assume a wrong cell count as it would be to assume a wrong empty voltage, I'd say: set the voltages to unset and force their configuration either by a GS wizzard or direct parameter setting.

I will make another commit to update the parameters tomorrow. Any thoughts on the actual monitoring algo?

**Comment by DonLakeFlyer** [comment]
Since you are removing parameters this is a breaking change to the parameter specification. When that happens you need to bump the SYS_PARAM_VER default. This signals QGC of the breaking change. QGC side is not fully implemented, but it's a good practice to get started with.

**Comment by LorenzMeier** [comment]
I'm happy to merge this if you bring the cell count back. There is no need to make it harder for users, and they almost never change the voltage thresholds, but they *have to* change the cell count for a number of airframes.
This is also important for smart batteries (which we already support) so that we can set the threshold correctly based on the reported number of cells. Not setting the cell count results in missing info.

**Comment by LorenzMeier** [comment]
Could you summarize again in one sentence what you feel the main improvement here is? This looks largely equivalent to what is in place, plus configurable percentages for warn and critical.

**Comment by orangelynx** [comment]
@DonLakeFlyer Thanks for the advice :+1: 
@LorenzMeier really? I searched through the project for source that actually uses the BAT_N_CELLS parameter, but couldn't find any except the estimation method which just uses it as a multiplication factor. I can of course bring it back if you need it - in this case would you like all the other parameters to be "per cell" as well?

The main improvement indeed is little, I opted for a simple model with maximum flexibility. The old voltage drop estimation was incomplete in the sense that it was neglecting hysteresis properties of the battery, leading to way to early battery warnings. This solution puts more focus on the coulomb-counting approach, avoiding this problem.

I will put more effort into developing a better monitoring system - this PR is mostly about making it work for everybody (by not forcing an incomplete battery model on them).

**Comment by DonLakeFlyer** [comment]
> I'm happy to merge this if you bring the cell count back.
> 
I agree since as Lorenz says, for most users they will just need to tweak cell counts. If would set the cell count to 0 as default. This can then trigger QGC to require that param be set as required. Then the code doesn’t run if cell count and mAh are not set. This prevents bad information if they haven’t been changed. I’d probably default mAh to 0 as well as the not set indicator as well instead of -1.

**Comment by orangelynx** [comment]
Yes, I see how having a cell parameter is more comfortable, however I am convinced that everything that makes configuration more comfortable should be part of the configuration in the GS, and that the px4 parameters should strictly back the algorithm which they are used in.

However if Lorenz says that he has some code for smart batteries that actually relies on the number of cells - that's a different story.

I have added 3 Commits.

Commit 1 fixes the parameters as suggested (i assume "not set" is achieved by setting to zero)
Commit 2 brings back the cell count parameter but the voltage parameters are still for total voltage
Commit 3 changes the voltage parameters to be per cell.

NOTE: bypassing coloumb-counting is still achieved by setting capacity to -1, not 0 and default cell count is still 1.

Please let me know if you want any further changes / manual optimization if you think that the optimizer won't take care of it.

**Comment by DonLakeFlyer** [comment]
> however I am convinced that everything that makes configuration more comfortable should be part of the configuration in the GS, and that the px4 parameters should strictly back the algorithm which they are used in.

That is way wrong. You should not assume that a ground station even exists and design your parameters for ease of use when all you have is parameter editing capability. Even if you have QGC, the majority of parameters have no custom ui on top of them and are only available through the parameter editor interface. If parameters (there are 500+) were not designed with end user usability in mind then I'd spend the rest of my life writing custom ui to make the flight stack usable. That is not feasible, nor does 100 different custom configuration screens provide end user usability. A parameter is an end user facing thing and requires end user facing thought put into it towards usability. You need to consider that end user to not be a developer. If we keep thinking that all parameters are just developer only facing knobs, the PX4 Flight Stack will never be more than something usable by rocket scientists only. That would suck.

**Comment by orangelynx** [comment]
Okay if that is the general design idea behind parameters I'll stick to that. My impression was just based on my first experience with the px4 flight stack - which would have been impossible to configure without the calibration wizards of the GS. The second thought was that userfriendly parameters lead to unnecessary calculations in the flight stack, which is just adds extra load to the RT system.. however i guess that was not a valid thought because the conversion technically just needs to be run once, thus not affecting the timing / performance of the RT system.

File: src/modules/commander/commander_params.c [hunk 3]
```diff
@@ -87,58 +87,86 @@ PARAM_DEFINE_FLOAT(TRIM_PITCH, 0.0f);
...
+ * Defines the number of cells the attached battery consists of.
  *
  * @group Battery Calibration
- * @unit V
+ * @unit S
+ * @min 1
+ * @max 10
  */
-PARAM_DEFINE_FLOAT(BAT_V_EMPTY, 3.4f);
+PARAM_DEFINE_INT32(BAT_N_CELLS, 1);
```
Comment (DonLakeFlyer) [inline src/modules/commander/commander_params.c]: Default to 0, so that setting is required?

File: src/modules/commander/commander_params.c [hunk 4]
```diff
@@ -87,58 +87,86 @@ PARAM_DEFINE_FLOAT(TRIM_PITCH, 0.0f);
...
  *
- * Defines the voltage where a single cell of the battery is considered full.
+ * Defines the voltage at which one battery cell is considered full.
  *
  * @group Battery Calibration
  * @unit V
+ * @min 0.0
  */
-PARAM_DEFINE_FLOAT(BAT_V_CHARGED, 4.2f);
+PARAM_DEFINE_FLOAT(BAT_V_FULL, 0.0f);
```
Comment (DonLakeFlyer) [inline src/modules/commander/commander_params.c]: Default to lipo cell value?

File: src/modules/commander/commander_params.c [hunk 5]
```diff
@@ -87,58 +87,86 @@ PARAM_DEFINE_FLOAT(TRIM_PITCH, 0.0f);
...
- * to maximum current ratio and assumes linearity.
+ * Defines the voltage at which one battery cell is considered low.
  *
  * @group Battery Calibration
  * @unit V
- * @min 0.0f
+ * @min 0.0
  */
-PARAM_DEFINE_FLOAT(BAT_V_LOAD_DROP, 0.07f);
+PARAM_DEFINE_FLOAT(BAT_V_LOW, 0.0f);
```
Comment (DonLakeFlyer) [inline src/modules/commander/commander_params.c]: Default to lipo cell value? Maybe add CELL to the name as well?

File: src/modules/commander/commander_params.c [hunk 6]
```diff
@@ -87,58 +87,86 @@ PARAM_DEFINE_FLOAT(TRIM_PITCH, 0.0f);
...
  *
  * @group Battery Calibration
- * @unit S
- * @min 1
- * @max 10
+ * @unit V
+ * @min 0.0
  */
-PARAM_DEFINE_INT32(BAT_N_CELLS, 3);
+PARAM_DEFINE_FLOAT(BAT_V_CRIT, 0.0f);
```
Comment (DonLakeFlyer) [inline src/modules/commander/commander_params.c]: Same thing here.

File: src/modules/commander/commander_params.c [hunk 7]
```diff
@@ -87,58 +87,86 @@ PARAM_DEFINE_FLOAT(TRIM_PITCH, 0.0f);
...
  *
  * Defines the capacity of the attached battery.
+ * Setting this value to -1 disables remaining battery estimation based on discharged capacity.
  *
  * @group Battery Calibration
- * @unit mA
+ * @unit mAh
  */
-PARAM_DEFINE_FLOAT(BAT_CAPACITY, -1.0f);
+PARAM_DEFINE_INT32(BAT_CAPACITY, -1);
```
Comment (DonLakeFlyer) [inline src/modules/commander/commander_params.c]: Use 0 a not set. Then set min to 0?

File: src/modules/commander/commander_params.c (hunk 3 above)
Comment (orangelynx) [inline src/modules/commander/commander_params.c]: ok

File: src/modules/commander/commander_params.c (hunk 4 above)
Comment (orangelynx) [inline src/modules/commander/commander_params.c]: could do, sure, however not everybody uses lipos.. and it would make the setting "preset"

File: src/modules/commander/commander_params.c (hunk 7 above)
Comment (orangelynx) [inline src/modules/commander/commander_params.c]: or maybe use 0 for "unknown" and -1 for "bypass charge based estimation"?

**Comment by orangelynx** [comment]
do not merge, latest commit seems not to be working, investigating now.

File: src/modules/commander/commander_params.c (hunk 4 above)
Comment (DonLakeFlyer) [inline src/modules/commander/commander_params.c]: That's got to be the 90% case though? Need feedback on that. You could add a comment in the long description that the default is for lipos. If not, lipo you'll need to adjust.

File: src/modules/commander/commander_params.c (hunk 7 above)
Comment (DonLakeFlyer) [inline src/modules/commander/commander_params.c]: What is the difference between the two? Seems both cases would bypass discharged capacity.

**Comment by orangelynx** [comment]
The latest commit fixes a some bugs, adjust the parameters as discussed and will also ensure  that battery parameters are updated when they are changed. It should be all good now, I will do some flight testing tomorrow and probably write a guide about checking voltage and current scaling parameters as well as tweaking the parameters. If I have time I might also look into updating QGC to deal with the new system.

**Comment by DonLakeFlyer** [comment]
That would be awesome!

**Comment by orangelynx** [comment]
Passed flight testing.

**Comment by bansiesta** [comment]
Reviewed. A little sad to see the voltage drop feature go because it worked for me but I agree that setting the capacity should actually be encouraged.

**Comment by orangelynx** [comment]
@banisiesta interesting to see that voltage drop estimation worked fine for you... how much drop did you encounter? When I come to revising the Power screen of QGC, I'll probably add a "fixed voltage drop" field there.

@LorenzMeier From my side its all clear for merge, or do you want to wait till a PR for QGC is ready as well?

**Comment by bansiesta** [comment]
@orangelynx I seemed to get less "false positives" if I increased the paramter BAT_V_LOAD_DROP. This is the only thing I remember.

**Comment by LorenzMeier** [comment]
What was the state of the matching QGC PR?

**Comment by DonLakeFlyer** [comment]
It’s not there. Has this pull gone through? If so I probably have enough time to put in a version check to disable battery config on new version.

**Comment by LorenzMeier** [comment]
No, its still pending and we might need a call to settle the discussion here.

**Comment by DonLakeFlyer** [comment]
Let me know on gitter if this ever goes through so I can do the version check.

**Comment by orangelynx** [comment]
Sorry currently have exams at ETH, so I have absolutely no spare time to work on this.. I will work on it full time starting september 1st.

**Comment by tumbili** [comment]
@orangelynx Did you find some time to work on this?

**Comment by LorenzMeier** [comment]
Closing for now, this needs a wider discussion and a fresh look.

**Comment by orangelynx** [comment]
@tumbili unfortunately not. Allthough I had zero problems so far with the modified firmware so far, I'm also doing an internship this semester and didn't find enough time to work through Qt to add the QGroundcontrol view for my revised system.

@LorenzMeier I'm always open to participate in another tackle at this. I hope I could at least show that the current implementation has its issues. I know that you see the ultimate solution in smart batteries, however I doubt that they will experience a breakthrough in the pixhawk user base any time soon.
